Docusaurus v2: improperly escaped routes on Windows

## 🐛 Bug Report

(A clear and concise description of what the bug is)

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

## To Reproduce

(Write your steps here:)

1. `git clone` and `cd` Docusaurus repo
2. `yarn install`
3. `cd website`
4. `yarn start`

## Expected behavior

The local server will start

## Actual Behavior

A blank page shows up at localhost:3000, and the terminal says: 

> Module not found: Can't resolve '@generateddocusaurus-plugin-content-blologMetadata.json' in 'C:\Users\nebrelbug\Documents\Coding\Docusaurus\website\.docusaurus'./.docusaurus/metadata.js

## Reproducible Demo

Follow the steps to reproduce

## What I think is the issue:

If you look in `website/.docusaurus/routes.js` and `website/.docusaurus/routes.js`, you can see that many of the routes look like this: `'C:\Users\nebrelbug\Documents\Coding\Docusaurus\docs\api-commands.md'` when they should look like: `C:\\Users\\nebrelbug\\Documents\\Coding\\Docusaurus\\docs\\api-commands.md`

I'm guessing this has to do with the fact that Windows uses `\` for paths

--- a/packages/docusaurus-utils/src/__tests__/index.test.js
+++ b/packages/docusaurus-utils/src/__tests__/index.test.js
@@ -13,9 +13,23 @@ import {
   idx,
   getSubFolder,
   normalizeUrl,
+  posixPath,
 } from '../index';
 
 describe('load utils', () => {
+  test('posixPath', () => {
+    const asserts = {
+      'c:/aaaa\\bbbb': 'c:/aaaa/bbbb',
+      'c:\\aaaa\\bbbb\\★': 'c:\\aaaa\\bbbb\\★',
+      '\\\\?\\c:\\aaaa\\bbbb': '\\\\?\\c:\\aaaa\\bbbb',
+      'c:\\aaaa\\bbbb': 'c:/aaaa/bbbb',
+      'foo\\bar': 'foo/bar',
+    };
+    Object.keys(asserts).forEach(file => {
+      expect(posixPath(file)).toBe(asserts[file]);
+    });
+  });
+
   test('fileToComponentName', () => {
     const asserts = {
       'index.md': 'MDIndex',

--- a/packages/docusaurus-utils/src/index.js
+++ b/packages/docusaurus-utils/src/index.js
@@ -49,6 +49,21 @@ function fileToComponentName(file) {
   return ext ? ext.toUpperCase() + str : str;
 }
 
+/**
+ * Convert Windows backslash paths to posix style paths. E.g: endi\\lie -> endi/lie
+ * @param {string} str windows backslash paths
+ * @returns {string} posix-style path
+ */
+function posixPath(str) {
+  const isExtendedLengthPath = /^\\\\\?\\/.test(str);
+  const hasNonAscii = /[^\u0000-\u0080]+/.test(str); // eslint-disable-line
+
+  if (isExtendedLengthPath || hasNonAscii) {
+    return str;
+  }
+  return str.replace(/\\/g, '/');
+}
+
 function generateChunkName(str, prefix) {
   const name = str === '/' ? 'index' : kebabHash(str);
   return prefix ? `${prefix}---${name}` : name;
@@ -153,4 +168,5 @@ module.exports = {
   idx,
   normalizeUrl,
   parse,
+  posixPath,
 };

--- a/packages/docusaurus/lib/server/load/plugins.js
+++ b/packages/docusaurus/lib/server/load/plugins.js
@@ -7,7 +7,7 @@
 
 const fs = require('fs-extra');
 const path = require('path');
-const {generate} = require('@docusaurus/utils');
+const {generate, posixPath} = require('@docusaurus/utils');
 
 module.exports = async function loadPlugins({pluginConfigs = [], context}) {
   // 1. Plugin Lifecycle - Initialization/Constructor
@@ -48,7 +48,10 @@ module.exports = async function loadPlugins({pluginConfigs = [], context}) {
         metadataFileName,
         JSON.stringify(content, null, 2),
       );
-      const contentPath = path.join('@generated', pluginContentPath);
+      // Note that we need to convert it into POSIX format because
+      // import XXXXX from '@generated\this-is\my\path' is incorrect
+      // import XXXXX from '@generated/this-is/my/path' is correct
+      const contentPath = posixPath(path.join('@generated', pluginContentPath));
 
       return {
         metadataKey,

--- a/packages/docusaurus/lib/server/load/routes.js
+++ b/packages/docusaurus/lib/server/load/routes.js
@@ -34,7 +34,8 @@ async function loadRoutes(pluginsRouteConfigs) {
     const importStr = isObj ? target.path : target;
     const queryStr = target.query ? `?${stringify(target.query)}` : '';
     const chunkName = generateChunkName(name || importStr, prefix);
-    return `() => import(/* webpackChunkName: '${chunkName}' */ '${importStr}${queryStr}')`;
+    const finalStr = JSON.stringify(importStr + queryStr);
+    return `() => import(/* webpackChunkName: '${chunkName}' */ ${finalStr})`;
   }
 
   function generateRouteCode(pluginRouteConfig) {

